monkey/infection_monkey/monkey.py

```diff
@@ -411,11 +413,20 @@ def _stop_relay(self):
         if self._relay and self._relay.is_alive():
             self._relay.stop()
 
-            while self._relay.is_alive() and not self._control_channel.should_agent_stop():
-                self._relay.join(timeout=5)
+            while self._relay.is_alive():
+                self._wait_for_clients_shutdown()
 
-            if self._control_channel.should_agent_stop():
-                self._relay.join(timeout=60)
+            self._wait_for_clients_killed()
+
+    @retry(times=5, exceptions=[IslandCommunicationError])
```

Comment:
Ideally, this would decorate just the `control_channel` function, not a function with other logic in it.